File: code/solutions-to-exercises/solutions-to-leetcode/unsorted/003_longest_substring_without_repeating_characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int max=0;
        int temp_max = 0;
        // 标记上一个重复字符的位置，map中小于此索引的记录是无效的
        int last_dup_loc = -1;
        unordered_map<char, int> map;
        for(int i=0; i<s.length();++i){
            char c = s[i];
            if(map.count(c) && map[c] > last_dup_loc){
                max = max > temp_max? max: temp_max;
                temp_max = i - map[c];
                last_dup_loc = map[c];
            }else{
                ++temp_max;
            }
            map[c] = i;
        }
        // 最后一次命中哈希表之后，没有再发现重复字符，则temp_max需要在此处理
        max = max > temp_max? max: temp_max;
        return max;
    }
};
/*
执行用时：12 ms, 在所有 C++ 提交中击败了79.20% 的用户
内存消耗：8.1 MB, 在所有 C++ 提交中击败了66.76% 的用户

备忘：
一趟遍历出结果，时间复杂度O(N)，
temp_max动态标记着当前的子字符串长度，
利用哈希表不断检查下一个字符是否在前面有重复，
重复了，则前面那个字符及其之前的字符都无效（map.count(c) && map[c] > last_dup_loc），
以所命中字符的下一个字符作为新的子字符串的起点（temp_max = i - map[c];）
*/

File: code/solutions-to-exercises/solutions-to-leetcode/specific_topic/algorithm_baisc/3_sliding_window/003_longest_substring_without_repeating_characters.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# created by pwz.wiki 2022
from _codingtools import performance_analyzer


@performance_analyzer
def solution0(s: str):
    """双指针::滑动窗口？
    没有搞清楚双指针、滑动窗口的区别，等再多做写题目后总结下

    本题思路：
    窗口不断向前，每次向前移动，判断是否引入重复字符
        若不引入，count值++，继续右移
        若引入，记录一次count值，窗口左边界收缩至无重复状态，继续右移
    tips：使用散列表判断是否引入重复字符
    时间O(n)，空间O(m) m=连续不重复字符数量
    """
    length = len(s)
    left = right = 0
    count = 0
    record = set()
    while right < length:
        if s[right] not in record:
            record.add(s[right])
            right += 1
        else:
            count = max(count, len(record))
            record.remove(s[left])
            left += 1
    count = max(count, len(record))
    return count


@performance_analyzer
def solution_official(s: str):
    """作者：LeetCode - Solution
链接：https://leetcode-cn.com/problems/longest-substring-without-repeating-characters/solution/wu-zhong-fu-zi-fu-de-zui-chang-zi-chuan-by-leetc-2/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。"""
    # 哈希集合，记录每个字符是否出现过
    occ = set()
    n = len(s)
    # 右指针，初始值为 -1，相当于我们在字符串的左边界的左侧，还没有开始移动
    rk, ans = -1, 0
    for i in range(n):
        if i != 0:
            # 左指针向右移动一格，移除一个字符
            occ.remove(s[i - 1])
        while rk + 1 < n and s[rk + 1] not in occ:
            # 不断地移动右指针
            occ.add(s[rk + 1])
            rk += 1
        # 第 i 到 rk 个字符是一个极长的无重复字符子串
        ans = max(ans, rk - i + 1)
    return ans


if __name__ == '__main__':
    print(solution0('lkajsdhflgkiuyaweuihgfbjhmadyyyyyyyyyyyyyyyyyyyyyysgfvuiljawheklfgvoiuyasghlkfjehgwljfgaskjdghfkjhwageuyrghialwsyr8723yr87gsjludgfuatw3ufyhsdgbkjfg2qu34grfkjhasdgvbkfjh'))
    print(solution_official('lkajsdhflgkiuyaweuihgfbjhmadyyyyyyyyyyyyyyyyyyyyyysgfvuiljawheklfgvoiuyasghlkfjehgwljfgaskjdghfkjhwageuyrghialwsyr8723yr87gsjludgfuatw3ufyhsdgbkjfg2qu34grfkjhasdgvbkfjh'))

"""
exec solution0():
cost: 0.07269999999999499ms
14

exec solution_offcial():
cost: 0.08129999999999943ms
14
"""